File: dyn_mix_cop/markov_dyn_mix_copula.r
# Libraries for copula and tail dependence analysis
# install.packages(c("copula", "mev"))
library(copula)
library(mev)

# --- 1. BASE COPULA DEFINITIONS ---

# Tail Copula: Gaussian (Asymptotically Independent)
d_tail_cop <- function(u_star, v_star, rho) {
  cop <- normalCopula(param = rho, dim = 2)
  return(dCopula(cbind(u_star, v_star), cop))
}

# Bulk Copula: Gumbel (Asymptotically Dependent)
d_bulk_cop <- function(u_star, v_star, alpha) {
  cop <- gumbelCopula(param = alpha, dim = 2)
  return(dCopula(cbind(u_star, v_star), cop))
}

# Dynamic Weighting Function: pi(u*, v*) = (u* * v*)^theta
pi_weight <- function(u_star, v_star, theta) {
  return((u_star * v_star)^theta)
}

# --- 2. THE LATENT MIXTURE DENSITY (c*) ---
# This defines the joint density in the non-uniform latent space u*.
# K is the normalization constant required to make this a valid density.
c_star_unnorm <- function(u_star, v_star, theta, rho, alpha) {
  w <- pi_weight(u_star, v_star, theta)
  # Blending: Tail copula dominates at high values of u* and v*
  dens <- w * d_tail_cop(u_star, v_star, rho) + (1 - w) * d_bulk_cop(u_star, v_star, alpha)
  return(dens)
}

# --- 3. MODEL PARAMETERS ---
theta_param <- 1.5   # Transition weight parameter
rho_param   <- 0.5   # Gaussian parameter (Expected Limit: eta = 0.75, chi = 0)
alpha_param <- 1.2   # Gumbel parameter (Bulk behavior: chi = 0.58)

# --- 4. GLOBAL NORMALIZATION & MAPPING (F_u*) ---
cat("Calculating Global Normalization Constant K and Marginal Splines...\n")

# Compute K: The volume of the unnormalized mixture over [0,1]^2
K_val <- integrate(function(v) {
  sapply(v, function(vi) {
    integrate(function(ui) c_star_unnorm(ui, vi, theta_param, rho_param, alpha_param), 0, 1)$value
  })
}, 0, 1)$value

# Marginal Latent Density f_u*(s): normalized by K
f_u_star <- function(s) {
  val <- integrate(function(v) c_star_unnorm(s, v, theta_param, rho_param, alpha_param), 0, 1)$value
  return(val / K_val)
}

# Pre-compute Marginal CDF F_u*(s) for the mapping: u* -> u
grid_size <- 50
u_star_grid <- seq(0.001, 0.999, length.out = grid_size)
F_grid <- sapply(u_star_grid, function(limit) {
  integrate(function(s) sapply(s, f_u_star), 0, limit)$value
})

# Create Splines for fast space-switching
# F_u_star_spline: Latent u* -> Uniform u
F_u_star_spline <- splinefun(u_star_grid, F_grid, method = "monoH.FC")
# F_u_star_inv_spline: Uniform u -> Latent u*
F_u_star_inv_spline <- splinefun(F_grid, u_star_grid, method = "monoH.FC")

# --- 5. THE TRANSITION KERNEL ---

# Conditional CDF in the latent space: P(v* <= target | u* = current)
# Note: K cancels out here because it's in both numerator and denominator.
cond_cdf_latent <- function(target_v_star, current_u_star) {
  num <- integrate(function(v) c_star_unnorm(current_u_star, v, theta_param, rho_param, alpha_param), 0, target_v_star)$value
  den <- f_u_star(current_u_star) * K_val
  return(num / den)
}

# Function to propagate the chain one step: u_t -> u_{t+1}
sample_next_step <- function(u_t) {
  # 1. Map current uniform state to latent space: u_t -> u*_t
  u_star_t <- F_u_star_inv_spline(u_t)
  
  # 2. Sample next state in latent space via inverse transform sampling
  w <- runif(1)
  u_star_next <- uniroot(function(v) cond_cdf_latent(v, u_star_t) - w, 
                         lower = 0, upper = 1, tol = 1e-6)$root
  
  # 3. Map back to uniform space to ensure stationarity: u*_next -> u_{t+1}
  u_next <- F_u_star_spline(u_star_next)
  return(u_next)
}

# --- 6. SIMULATION ---
n_steps <- 200
u_chain <- numeric(n_steps)
u_chain[1] <- runif(1) # Start with a random uniform value

cat("Simulating Univariate Markov Chain...\n")
for(t in 1:(n_steps-1)) {
  u_chain[t+1] <- sample_next_step(u_chain[t])
  if(t %% 100 == 0) cat("Step", t, "completed...\n")
}

x_chain <- egpd::qegpd(u_chain, kappa = 2, sigma = 1, xi = 0.3)

# --- 7. VALIDATION ---
# Create consecutive pairs to check dependence
u_pairs <- cbind(u_chain[1:(n_steps-1)], u_chain[2:n_steps])

par(mfrow=c(1,2))
plot(u_chain, type="l", col="steelblue", main="Markov Chain Trace (u)", ylab="u_t")
plot(u_pairs, pch=20, col=rgb(0,0,0,0.2), main="Transition Copula (u_t, u_{t+1})",
     xlab="u_t", ylab="u_{t+1}")

# Check Tail Dependence (using mev package)
cat("\n--- Tail Dependence Analysis ---\n")
td <- mev::taildep(u_pairs, u = seq(0.7, 0.98, by = 0.01))

# Theoretical Values based on Gaussian Tail (rho=0.5) and Gumbel Bulk (alpha=2)
cat("Theoretical Tail Limit (r -> 1):\n")
cat("Expected chi: 0.00\n")
cat("Expected eta: ", (1 + rho_param)/2, "\n")

File: dyn_mix_cop/markov_dyn_mix_copula.jl
# markov_chain.jl
using QuadGK, Roots, Distributions, Interpolations
using DataFrames
using Random

# ------------------------------------------------------------------
# 1. COPULA DENSITIES
# ------------------------------------------------------------------

"""Bivariate Gaussian copula density via the conditional approach."""
function d_gaussian_copula(u, v, rho)
    # Transform to normal scores
    x = quantile(Normal(), u)
    y = quantile(Normal(), v)
    # Bivariate normal density / product of marginal normals
    z = (x^2 - 2*rho*x*y + y^2) / (1 - rho^2)
    return exp(-z/2 + (x^2 + y^2)/2) / sqrt(1 - rho^2)
end

"""Bivariate Gumbel copula density."""
function d_gumbel_copula(u, v, alpha)
    # Avoid boundary issues
    u = clamp(u, 1e-10, 1 - 1e-10)
    v = clamp(v, 1e-10, 1 - 1e-10)
    lu = -log(u); lv = -log(v)
    s  = (lu^alpha + lv^alpha)^(1/alpha)
    # Log-scale for numerical stability
    log_c = (alpha - 1)*log(lu) + (alpha - 1)*log(lv) -
            (lu^alpha + lv^alpha)^(1/alpha) +
            log(s^(1/alpha) + alpha - 1) -
            (1 + 1/alpha)*log(lu^alpha + lv^alpha) -
            log(u) - log(v)
    return exp(log_c)
end

# ------------------------------------------------------------------
# 2. MIXTURE DENSITY
# ------------------------------------------------------------------

pi_weight(u, v, theta) = (u * v)^theta

function c_star_unnorm(u, v, theta, rho, alpha)
    w   = pi_weight(u, v, theta)
    d_t = d_gaussian_copula(u, v, rho)
    d_b = d_gumbel_copula(u, v, alpha)
    return w * d_t + (1 - w) * d_b
end

# ------------------------------------------------------------------
# 3. PRECOMPUTATION (K, marginal CDF spline)
# ------------------------------------------------------------------

struct ModelPrecomp
    K          :: Float64
    u_star_grid:: Vector{Float64}
    F_grid     :: Vector{Float64}
    theta      :: Float64
    rho        :: Float64
    alpha      :: Float64
    F_spline   :: Any   # add spline object
    Finv_spline:: Any   # inverse spline
end

function build_precomp(theta, rho, alpha; grid_size=100)

    K, _ = quadgk(v -> quadgk(u -> c_star_unnorm(u, v, theta, rho, alpha),
                               0.001, 0.999)[1],
                   0.001, 0.999, rtol=1e-5)

    f_u_star(s) = quadgk(v -> c_star_unnorm(s, v, theta, rho, alpha),
                          0.001, 0.999)[1] / K

    u_grid = range(0.001, 0.999, length=grid_size) |> collect
    F_vals = similar(u_grid)

    for (i, lim) in enumerate(u_grid)
        F_vals[i], _ = quadgk(f_u_star, 0.001, lim, rtol=1e-5)
    end

    F_spline = LinearInterpolation(u_grid, F_vals, extrapolation_bc=Flat())
    Finv_spline = LinearInterpolation(F_vals, u_grid, extrapolation_bc=Flat())

    return ModelPrecomp(K, u_grid, F_vals, theta, rho, alpha, F_spline, Finv_spline)
end

# ------------------------------------------------------------------
# 4. SPLINE HELPERS  (linear interp fallback — swap for DataInterpolations)
# ------------------------------------------------------------------

# function interp1d(x, y, xi)
#     # Simple monotone linear interpolation (replace with spline if needed)
#     i = searchsortedfirst(x, xi) - 1
#     i = clamp(i, 1, length(x)-1)
#     t = (xi - x[i]) / (x[i+1] - x[i])
#     return y[i] + t * (y[i+1] - y[i])
# end

# F_u_star(p::ModelPrecomp, s)    = interp1d(p.u_star_grid, p.F_grid, s)
# F_u_star_inv(p::ModelPrecomp, u) = interp1d(p.F_grid, p.u_star_grid, u)

F_u_star(p::ModelPrecomp, s) = p.F_spline(s)
F_u_star_inv(p::ModelPrecomp, u) = p.Finv_spline(u)

# ------------------------------------------------------------------
# 5. TRANSITION KERNEL
# ------------------------------------------------------------------

function cond_cdf_latent(target_v, current_u_star, p::ModelPrecomp)
    num, _ = quadgk(v -> c_star_unnorm(current_u_star, v, p.theta, p.rho, p.alpha),
                     0.001, target_v, rtol=1e-6)
    den, _ = quadgk(v -> c_star_unnorm(current_u_star, v, p.theta, p.rho, p.alpha),
                     0.001, 0.999, rtol=1e-6)
    return num / den  # K cancels
end

function sample_next_step(u_t, p::ModelPrecomp)
    u_star_t    = F_u_star_inv(p, u_t)
    w           = rand()
    # Inverse-transform: find v* such that CDF(v*) = w
    f(v)        = cond_cdf_latent(v, u_star_t, p) - w
    u_star_next = find_zero(f, (0.001, 0.999), Bisection())
    return F_u_star(p, u_star_next)
end

# ------------------------------------------------------------------
# 6. SIMULATION ENTRY POINT
# ------------------------------------------------------------------

function simulate_chain(n_steps, theta, rho, alpha; seed=nothing)
    isnothing(seed) || Random.seed!(seed)
    println("Building precomputation...")
    p = build_precomp(theta, rho, alpha)
    println("  K = $(p.K)")

    chain = Vector{Float64}(undef, n_steps)
    chain[1] = rand()
    println("Simulating $n_steps steps...")
    for t in 1:(n_steps-1)
        chain[t+1] = sample_next_step(chain[t], p)
        t % 100 == 0 && println("  Step $t done")
    end
    return chain
end